services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowSyncSpec.groovy

```diff
@@ -0,0 +1,101 @@
+package org.openkilda.functionaltests.spec.northbound.flows
+
+import static org.junit.Assume.assumeTrue
+import static org.openkilda.testing.Constants.RULES_DELETION_TIME
+import static org.openkilda.testing.Constants.RULES_INSTALLATION_TIME
+
+import org.openkilda.functionaltests.BaseSpecification
+import org.openkilda.functionaltests.helpers.FlowHelper
+import org.openkilda.functionaltests.helpers.PathHelper
+import org.openkilda.functionaltests.helpers.Wrappers
+import org.openkilda.messaging.model.SwitchId
+import org.openkilda.testing.Constants.DefaultRule
+import org.openkilda.testing.model.topology.TopologyDefinition
+import org.openkilda.testing.model.topology.TopologyDefinition.Switch
+import org.openkilda.testing.service.database.Database
+import org.openkilda.testing.service.northbound.NorthboundService
+import org.openkilda.testing.tools.IslUtils
+
+import groovy.time.TimeCategory
+import org.springframework.beans.factory.annotation.Autowired
+import spock.lang.Shared
+
+import java.util.concurrent.TimeUnit
+
+class FlowSyncSpec extends BaseSpecification {
+
+    @Autowired
+    TopologyDefinition topology
+    @Autowired
+    FlowHelper flowHelper
+    @Autowired
+    NorthboundService northboundService
+    @Autowired
+    PathHelper pathHelper
+    @Autowired
+    IslUtils islUtils
+    @Autowired
+    Database db
+
+    @Shared
+    int flowRulesCount = 2
+
+    def "Able to synchronize a flow (install missing flow rules, reinstall existing) without rerouting"() {
+        given: "A flow with one deleted rule on each switch"
+        def switches = topology.getActiveSwitches()
+        def allLinks = northboundService.getAllLinks()
+        def (Switch srcSwitch, Switch dstSwitch) = [switches, switches].combinations()
+                .findAll { src, dst -> src != dst }.find { Switch src, Switch dst ->
+            allLinks.every { link ->
+                def switchIds = link.path*.switchId
+                !(switchIds.contains(src.dpId) && switchIds.contains(dst.dpId))
+            }
+        } ?: assumeTrue("No suiting switches found", false)
+
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flowHelper.addFlow(flow)
+        def flowPath = PathHelper.convert(northboundService.getFlowPath(flow.id))
+
+        def involvedSwitches = pathHelper.getInvolvedSwitches(flow.id)
+        def ruleToDelete = getFlowRules(involvedSwitches.first().dpId).first().cookie
+        involvedSwitches.each { sw ->
+            northboundService.deleteSwitchRules(sw.dpId, ruleToDelete)
+            Wrappers.wait(RULES_DELETION_TIME) {
+                assert getFlowRules(sw.dpId).size() == flowRulesCount - 1
+            }
+        }
+
+        when: "Synchronize the flow"
+        TimeUnit.SECONDS.sleep(1)  // Let's wait a bit to give rules at least 1 second to be present on switches
```

Comment:
I'm not sure why this is required. You can use nanoseconds field if accuracy it too low for the test, but I think even with seconds this should work fine. I think you used to experience problems because of not putting second verification inside 'wait'

Reply:
done.